pymc3/distributions/distribution.py

```diff
@@ -134,40 +321,147 @@ def __new__(cls, name, *args, **kwargs):
                 "for a standalone distribution."
             )
 
-        rng = kwargs.pop("rng", None)
-
-        if rng is None:
-            rng = model.default_rng
-
         if not isinstance(name, string_types):
             raise TypeError(f"Name needs to be a string but got: {name}")
 
-        data = kwargs.pop("observed", None)
+        if rng is None:
+            rng = model.default_rng
 
-        total_size = kwargs.pop("total_size", None)
+        if dims is not None and "shape" in kwargs:
+            raise ValueError(
+                f"Passing both `dims` ({dims}) and `shape` ({kwargs['shape']}) is not supported!"
+            )
+        if dims is not None and "size" in kwargs:
+            raise ValueError(
+                f"Passing both `dims` ({dims}) and `size` ({kwargs['size']}) is not supported!"
+            )
+        dims = _convert_dims(dims)
 
-        dims = kwargs.pop("dims", None)
+        # Create the RV without specifying testval, because the testval may have a shape
+        # that only matches after replicating with a size implied by dims (see below).
+        rv_out = cls.dist(*args, rng=rng, testval=None, **kwargs)
+        ndim_implied = rv_out.ndim
+        resize_shape = None
 
-        if "shape" in kwargs:
-            raise DeprecationWarning("The `shape` keyword is deprecated; use `size`.")
+        # `dims` are only available with this API, because `.dist()` can be used
+        # without a modelcontext and dims are not tracked at the Aesara level.
+        if dims is not None:
+            ndim_resize, resize_shape, dims = _resize_from_dims(dims, ndim_implied, model)
+        elif observed is not None:
```

Comment:
Should a variable allow `dims` and `observed`, and if so, are we checking whether they match anywhere?

Reply:
Not sure about the checking but we definitely should allow both and we already have several examples in the docs doing so too

Reply:
The changes from this PR make a few more restrictions on the compatibility of `observed`. Specifically the RV is no longer automatically resized based on `observed`, because that could conflict with a dimensionality parametrization through either of the other kwargs.

Reply:
I asked because of that `elif`, if there are `dims`, `_resize_from_observed` won't be called. Is this safe?

Reply:
Also do we have new tests for the "new" restrictions on observed?

Reply:
This is a matter of having a single source of truth -- in this case the `dims` kwarg takes priority over the `observed`, because it belongs to the family of dimensionality kwargs. `Model.make_obs_var` raises an error if the dimensionality does not match:
```
if data.ndim != rv_var.ndim:
      raise ShapeError(
          "Dimensionality of data and RV don't match.", actual=data.ndim, expected=rv_var.ndim
      )
```